File: ios/Asistan/Services/APIService.swift
import Foundation
import Supabase

/// User-facing error carrying a Turkish message from the backend.
nonisolated struct APIError: LocalizedError {
    let message: String
    var errorDescription: String? { message }
}

private nonisolated struct ErrorBody: Decodable, Sendable {
    let error: String?
    let message: String?
}

/// Thin wrapper over Supabase Edge Functions. Decodes responses and surfaces
/// backend Turkish error messages instead of opaque HTTP failures.
nonisolated enum APIService {
    private static func invoke<Body: Encodable & Sendable, T: Decodable & Sendable>(
        _ name: String,
        body: Body
    ) async throws -> T {
        do {
            let result: T = try await supabase.functions.invoke(name, options: .init(body: body))
            return result
        } catch let FunctionsError.httpError(code, data) {
            if let parsed = try? JSONDecoder().decode(ErrorBody.self, from: data), let msg = parsed.message {
                throw APIError(message: msg)
            }
            throw APIError(message: "İşlem tamamlanamadı (\(code)). Lütfen tekrar deneyin.")
        } catch let error as APIError {
            throw error
        } catch {
            throw APIError(message: "Bağlantı hatası. İnternet bağlantınızı kontrol edin.")
        }
    }

    // MARK: Auth / Profile

    nonisolated struct RegisterRequest: Encodable, Sendable {
        let action = "register"
        let fullName: String
        let email: String
        let phone: String
        let password: String
    }

    static func register(fullName: String, email: String, phone: String, password: String) async throws {
        let _: RegisterResponse = try await invoke(
            "auth-client",
            body: RegisterRequest(fullName: fullName, email: email, phone: phone, password: password)
        )
    }

    nonisolated struct GetProfileRequest: Encodable, Sendable {
        let action = "get"
        let metadata: Metadata?
        nonisolated struct Metadata: Encodable, Sendable {
            let fullName: String?
            let phone: String?
        }
    }

    static func getProfile(fullName: String? = nil, phone: String? = nil) async throws -> ClientUser? {
        let res: ClientUserResponse = try await invoke(
            "auth-client",
            body: GetProfileRequest(metadata: .init(fullName: fullName, phone: phone))
        )
        return res.clientUser
    }

    nonisolated struct UpdateProfileRequest: Encodable, Sendable {
        let action = "update"
        let fullName: String?
        let phone: String?
        let address: String?
        let city: String?
        let locationLat: Double?
        let locationLng: Double?
    }

    static func updateProfile(
        fullName: String? = nil,
        phone: String? = nil,
        address: String? = nil,
        city: String? = nil,
        locationLat: Double? = nil,
        locationLng: Double? = nil
    ) async throws -> ClientUser? {
        let res: ClientUserResponse = try await invoke(
            "auth-client",
            body: UpdateProfileRequest(
                fullName: fullName, phone: phone, address: address,
                city: city, locationLat: locationLat, locationLng: locationLng
            )
        )
        return res.clientUser
    }

    // MARK: Catalog

    nonisolated struct DiscoveryRequest: Encodable, Sendable {
        let action = "discovery"
        let lat: Double?
        let lng: Double?
    }

    static func discovery(lat: Double?, lng: Double?) async throws -> DiscoveryResponse {
        try await invoke("catalog", body: DiscoveryRequest(lat: lat, lng: lng))
    }

    nonisolated struct SearchFilters: Encodable, Sendable {
        var specialty: String?
        var service: String?
        var maxDistanceKm: Double?
        var minRating: Double?
        var availableToday: Bool?
        var maxPrice: Double?
        var clinicName: String?
        var doctorName: String?
    }

    nonisolated struct SearchRequest: Encodable, Sendable {
        let action = "search"
        let lat: Double?
        let lng: Double?
        let query: String?
        let filters: SearchFilters
        let sort: String
    }

    static func search(lat: Double?, lng: Double?, query: String?, filters: SearchFilters, sort: String) async throws -> [Provider] {
        let res: SearchResponse = try await invoke(
            "catalog",
            body: SearchRequest(lat: lat, lng: lng, query: query, filters: filters, sort: sort)
        )
        return res.providers
    }

    nonisolated struct ClinicRequest: Encodable, Sendable {
        let action = "clinic"
        let businessId: String
        let lat: Double?
        let lng: Double?
    }

    static func clinic(businessId: String, lat: Double?, lng: Double?) async throws -> ClinicResponse {
        try await invoke("catalog", body: ClinicRequest(businessId: businessId, lat: lat, lng: lng))
    }

    nonisolated struct DoctorRequest: Encodable, Sendable {
        let action = "doctor"
        let staffId: String
        let lat: Double?
        let lng: Double?
    }

    static func doctor(staffId: String, lat: Double?, lng: Double?) async throws -> DoctorResponse {
        try await invoke("catalog", body: DoctorRequest(staffId: staffId, lat: lat, lng: lng))
    }

    nonisolated struct SlotsRequest: Encodable, Sendable {
        let action = "slots"
        let staffId: String
        let serviceId: String
        let date: String
    }

    static func slots(staffId: String, serviceId: String, date: String) async throws -> [SlotItem] {
        let res: SlotsResponse = try await invoke(
            "catalog",
            body: SlotsRequest(staffId: staffId, serviceId: serviceId, date: date)
        )
        return res.slots
    }

    // MARK: Booking

    nonisolated struct BookRequest: Encodable, Sendable {
        let action = "book"
        let businessId: String
        let staffId: String
        let serviceId: String
        let date: String
        let startTime: String
        let note: String?
        let contactPhone: String?
    }

    static func book(
        businessId: String, staffId: String, serviceId: String,
        date: String, startTime: String, note: String?, contactPhone: String?
    ) async throws -> BookingResponse {
        try await invoke(
            "booking",
            body: BookRequest(
                businessId: businessId, staffId: staffId, serviceId: serviceId,
                date: date, startTime: startTime, note: note, contactPhone: contactPhone
            )
        )
    }

    nonisolated struct ListRequest: Encodable, Sendable { let action = "list" }

    static func myAppointments() async throws -> [AppointmentRow] {
        let res: AppointmentsResponse = try await invoke("booking", body: ListRequest())
        return res.appointments
    }

    nonisolated struct CancelRequest: Encodable, Sendable {
        let action = "cancel"
        let appointmentId: String
    }

    static func cancel(appointmentId: String) async throws {
        let _: OkResponse = try await invoke("booking", body: CancelRequest(appointmentId: appointmentId))
    }

    // MARK: Engagement

    nonisolated struct NotificationsRequest: Encodable, Sendable { let action = "notifications" }

    static func notifications() async throws -> NotificationsResponse {
        try await invoke("engagement", body: NotificationsRequest())
    }

    nonisolated struct MarkReadRequest: Encodable, Sendable {
        let action = "mark_read"
        let id: String
    }

    static func markRead(id: String) async throws {
        let _: OkResponse = try await invoke("engagement", body: MarkReadRequest(id: id))
    }

    nonisolated struct MarkAllReadRequest: Encodable, Sendable { let action = "mark_all_read" }

    static func markAllRead() async throws {
        let _: OkResponse = try await invoke("engagement", body: MarkAllReadRequest())
    }

    nonisolated struct ReviewRequest: Encodable, Sendable {
        let action = "review"
        let appointmentId: String
        let rating: Int
        let comment: String?
        let serviceQuality: Int?
        let waitingTime: Int?
        let communication: Int?
    }

    static func submitReview(
        appointmentId: String, rating: Int, comment: String?,
        serviceQuality: Int?, waitingTime: Int?, communication: Int?
    ) async throws {
        let _: OkResponse = try await invoke(
            "engagement",
            body: ReviewRequest(
                appointmentId: appointmentId, rating: rating, comment: comment,
                serviceQuality: serviceQuality, waitingTime: waitingTime, communication: communication
            )
        )
    }
}

File: android/app/src/main/java/com/rork/asistanrezervasyonandroid/data/ApiService.kt
package com.rork.asistanrezervasyonandroid.data

import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.contentType
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

@Serializable
private data class ErrorBody(
    val error: String? = null,
    val message: String? = null,
)

/** Search filters mirroring the iOS SearchFilters payload. */
data class SearchFilters(
    val specialty: String? = null,
    val service: String? = null,
    val maxDistanceKm: Double? = null,
    val minRating: Double? = null,
    val availableToday: Boolean? = null,
    val maxPrice: Double? = null,
    val clinicName: String? = null,
    val doctorName: String? = null,
) {
    val activeCount: Int
        get() {
            var n = 0
            if (specialty != null) n++
            if (service != null) n++
            if (maxDistanceKm != null) n++
            if (minRating != null) n++
            if (availableToday == true) n++
            if (maxPrice != null) n++
            return n
        }

    fun toJson(): JsonObject = buildJsonObject {
        specialty?.let { put("specialty", it) }
        service?.let { put("service", it) }
        maxDistanceKm?.let { put("maxDistanceKm", it) }
        minRating?.let { put("minRating", it) }
        availableToday?.let { put("availableToday", it) }
        maxPrice?.let { put("maxPrice", it) }
        clinicName?.let { put("clinicName", it) }
        doctorName?.let { put("doctorName", it) }
    }
}

/**
 * Thin wrapper over Supabase Edge Functions. Decodes responses and surfaces
 * backend Turkish error messages instead of opaque HTTP failures.
 */
object ApiService {

    private suspend inline fun <reified T> invoke(name: String, body: JsonObject): T {
        val token = AppGraph.auth.validAccessToken() ?: Env.supabaseAnonKey
        val text: String
        val statusCode: Int
        try {
            val res = AppGraph.http.post("${Env.supabaseUrl}/functions/v1/$name") {
                header("apikey", Env.supabaseAnonKey)
                header("Authorization", "Bearer $token")
                contentType(ContentType.Application.Json)
                setBody(body.toString())
            }
            statusCode = res.status.value
            text = res.bodyAsText()
        } catch (e: ApiException) {
            throw e
        } catch (_: Exception) {
            throw ApiException("Bağlantı hatası. İnternet bağlantınızı kontrol edin.")
        }
        if (statusCode !in 200..299) {
            val parsed = runCatching { AppJson.decodeFromString<ErrorBody>(text) }.getOrNull()
            throw ApiException(
                parsed?.message ?: "İşlem tamamlanamadı ($statusCode). Lütfen tekrar deneyin."
            )
        }
        return try {
            AppJson.decodeFromString<T>(text)
        } catch (_: Exception) {
            throw ApiException("Beklenmeyen yanıt alındı. Lütfen tekrar deneyin.")
        }
    }

    // MARK: Auth / Profile

    suspend fun register(fullName: String, email: String, phone: String, password: String) {
        invoke<RegisterResponse>("auth-client", buildJsonObject {
            put("action", "register")
            put("fullName", fullName)
            put("email", email)
            put("phone", phone)
            put("password", password)
        })
    }

    suspend fun getProfile(fullName: String? = null, phone: String? = null): ClientUser? {
        val res = invoke<ClientUserResponse>("auth-client", buildJsonObject {
            put("action", "get")
            put("metadata", buildJsonObject {
                fullName?.let { put("fullName", it) }
                phone?.let { put("phone", it) }
            })
        })
        return res.clientUser
    }

    suspend fun updateProfile(
        fullName: String? = null,
        phone: String? = null,
        address: String? = null,
        city: String? = null,
        locationLat: Double? = null,
        locationLng: Double? = null,
    ): ClientUser? {
        val res = invoke<ClientUserResponse>("auth-client", buildJsonObject {
            put("action", "update")
            fullName?.let { put("fullName", it) }
            phone?.let { put("phone", it) }
            address?.let { put("address", it) }
            city?.let { put("city", it) }
            locationLat?.let { put("locationLat", it) }
            locationLng?.let { put("locationLng", it) }
        })
        return res.clientUser
    }

    // MARK: Catalog

    suspend fun discovery(lat: Double?, lng: Double?): DiscoveryResponse =
        invoke("catalog", buildJsonObject {
            put("action", "discovery")
            lat?.let { put("lat", it) }
            lng?.let { put("lng", it) }
        })

    suspend fun search(
        lat: Double?,
        lng: Double?,
        query: String?,
        filters: SearchFilters,
        sort: String,
    ): List<Provider> {
        val res = invoke<SearchResponse>("catalog", buildJsonObject {
            put("action", "search")
            lat?.let { put("lat", it) }
            lng?.let { put("lng", it) }
            query?.let { put("query", it) }
            put("filters", filters.toJson())
            put("sort", sort)
        })
        return res.providers
    }

    suspend fun clinic(businessId: String, lat: Double?, lng: Double?): ClinicResponse =
        invoke("catalog", buildJsonObject {
            put("action", "clinic")
            put("businessId", businessId)
            lat?.let { put("lat", it) }
            lng?.let { put("lng", it) }
        })

    suspend fun doctor(staffId: String, lat: Double?, lng: Double?): DoctorResponse =
        invoke("catalog", buildJsonObject {
            put("action", "doctor")
            put("staffId", staffId)
            lat?.let { put("lat", it) }
            lng?.let { put("lng", it) }
        })

    suspend fun slots(staffId: String, serviceId: String, date: String): List<SlotItem> {
        val res = invoke<SlotsResponse>("catalog", buildJsonObject {
            put("action", "slots")
            put("staffId", staffId)
            put("serviceId", serviceId)
            put("date", date)
        })
        return res.slots
    }

    // MARK: Booking

    suspend fun book(
        businessId: String,
        staffId: String,
        serviceId: String,
        date: String,
        startTime: String,
        note: String?,
        contactPhone: String?,
    ): BookingResponse = invoke("booking", buildJsonObject {
        put("action", "book")
        put("businessId", businessId)
        put("staffId", staffId)
        put("serviceId", serviceId)
        put("date", date)
        put("startTime", startTime)
        note?.let { put("note", it) }
        contactPhone?.let { put("contactPhone", it) }
    })

    suspend fun myAppointments(): List<AppointmentRow> {
        val res = invoke<AppointmentsResponse>("booking", buildJsonObject { put("action", "list") })
        return res.appointments
    }

    suspend fun cancel(appointmentId: String) {
        invoke<OkResponse>("booking", buildJsonObject {
            put("action", "cancel")
            put("appointmentId", appointmentId)
        })
    }

    // MARK: Engagement

    suspend fun notifications(): NotificationsResponse =
        invoke("engagement", buildJsonObject { put("action", "notifications") })

    suspend fun markRead(id: String) {
        invoke<OkResponse>("engagement", buildJsonObject {
            put("action", "mark_read")
            put("id", id)
        })
    }

    suspend fun markAllRead() {
        invoke<OkResponse>("engagement", buildJsonObject { put("action", "mark_all_read") })
    }

    suspend fun submitReview(
        appointmentId: String,
        rating: Int,
        comment: String?,
        serviceQuality: Int?,
        waitingTime: Int?,
        communication: Int?,
    ) {
        invoke<OkResponse>("engagement", buildJsonObject {
            put("action", "review")
            put("appointmentId", appointmentId)
            put("rating", rating)
            comment?.let { put("comment", it) }
            serviceQuality?.let { put("serviceQuality", it) }
            waitingTime?.let { put("waitingTime", it) }
            communication?.let { put("communication", it) }
        })
    }
}
